features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/base64/Base64Serializer.kt

```diff
@@ -127,13 +139,35 @@ internal class Base64Serializer private constructor(
             return
         }
 
-        val base64String = convertBitmapToBase64(
-            byteArray = byteArray,
-            bitmap = bitmap,
-            shouldCacheBitmap = shouldCacheBitmap
-        ).toByteArray(Charsets.UTF_8)
-        val resourceId = md5HashGenerator.generate(byteArray)?.toByteArray(Charsets.UTF_8)
-        val cacheData = CacheData(base64String, resourceId)
+        val resourceId = md5HashGenerator.generate(byteArray)
+        var base64String = "".toByteArray(Charsets.UTF_8)
+        var cacheData = CacheData(base64String, resourceId?.toByteArray(Charsets.UTF_8))
+
+        if (shouldCacheBitmap) {
+            bitmapPool?.put(bitmap)
+        }
+
+        if (RESOURCE_ENDPOINT_FEATURE_FLAG) {
+            if (resourceId == null) {
+                // resourceId is mandatory for resource endpoint
+                base64SerializerCallback.onReady()
+                return
+            }
+
+            if (!previouslySentResources.contains(resourceId)) {
+                previouslySentResources.add(resourceId)
+                recordedDataQueueHandler.addResourceItem(
+                    identifier = resourceId,
+                    resourceData = byteArray,
+                    applicationId = applicationId
+                )
```

Comment:
this seems like a design flaw: imo the only job of serializer is to serialize, why do we need the queue handler here? Also by having this we are injecting `applicationId` into the serializer constructor, which looks strange (same reason: serializer should just serialize and return result), and to support this we need to pass `applicationId` through the mappers chain.

Reply:
We need to refactor this class and to split up the responsibilities. In the next pr I'll in any case rename it and also see what can be extracted out into separate components (since most of the logic in the class concerns getting the bitmaps). Regarding the applicationId let's discuss and see if there's a better way, because the `addResourceItem` is happening asynchronously

Reply:
Let's do this properly in the next PR, I do agree with @0xnm that we need to split and provide single responsibilities here.